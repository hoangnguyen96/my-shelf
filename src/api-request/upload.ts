import { MESSAGES } from "@app/constants";

export const generateImageUpload = async (file: File) => {
  try {
    const formData = new FormData();
    formData.append("image", file);

    const response = await fetch(
      `https://api.imgbb.com/1/upload?key=${process.env.NEXT_PUBLIC_IMGBB_API_KEY}`,
      {
        method: "POST",
        body: formData,
      }
    );

    if (!response.ok) {
      throw new Error(MESSAGES.UPLOAD_FAILED);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    throw new Error(MESSAGES.ERROR_UPLOAD);
  }
};
