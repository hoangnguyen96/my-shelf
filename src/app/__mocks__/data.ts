export const DATA_BOOKS = [
  {
    title: "Don't Make Me Think",
    author: "Steve Krug",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/QbQgVtG/book1.png",
    description: "description will update later",
    status: false,
    publicationYear: 2000,
    rating: 4.5,
    createdDate: "3 Sep 2024 9:07 PM",
    edition: "first",
    id: "1",
  },
  {
    title: "Rich Dad Poor Dad",
    author: "Robert T. Kiyosaki",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/9qHTtQF/book2.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1997,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:47 PM",
    edition: "second",
    id: "2",
  },
  {
    title: "The Design of Everyday Things",
    author: "Don Norman",
    category: "TextBook",
    imageUrl: "https://i.ibb.co/LnSBcCM/book5.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1988,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:51 PM",
    edition: "fourth",
    id: "3",
  },
];

export const DATA_USER = [
  {
    username: "admin",
    email: "admin@gmail.com",
    password: "$2a$10$MwSZ9vWSX7DVJwJ4SF5Mzuj2zMFjY52XM9L11gGG0j79gybowaLfq",
    isAdmin: true,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/RHMqQGr/man-1.png",
    favorites: ["3", "2", "1", "13", "5"],
    shelfBooks: ["1", "6"],
    userId: "3733403",
    id: "1",
  },
  {
    username: "hung",
    email: "hung1@gmail.com",
    password: "$2a$10$Mt6ruwOzt3nzX1.HpSt.OOzut.teXmLE74WclmkP.Nn.gzVQrlqri",
    isAdmin: false,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/88X1WfZ/avatar-default.png",
    favorites: ["1"],
    shelfBooks: ["2", "3", "1"],
    userId: "8857134",
    id: "2",
  },
];
