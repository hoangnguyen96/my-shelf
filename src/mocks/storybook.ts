export const mockRouter = {
  pathname: "/",
  route: "/",
  query: {},
  asPath: "/",
  basePath: "",
  push: () => Promise.resolve(true),
  replace: () => Promise.resolve(true),
  reload: () => {},
  back: () => {},
  prefetch: () => Promise.resolve(),
  beforePopState: () => {},
  isFallback: false,
  isLocaleDomain: false,
  isReady: true,
  isPreview: false,
  forward: () => {},
  events: {
    on: () => {},
    off: () => {},
    emit: () => {},
  },
};

export const mockSession = {
  user: {
    isAdmin: true,
    email: "admin@gmail.com",
    id: "3733403",
    name: "admin",
    image: "https://i.ibb.co/RHMqQGr/man-1.png",
  },
  expires: "2024-12-31T23:59:59.999Z",
};
