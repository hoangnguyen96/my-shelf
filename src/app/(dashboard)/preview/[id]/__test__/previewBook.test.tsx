import { act, render } from "@testing-library/react";
import { useSession } from "next-auth/react";
import { getBookById, getUserById } from "@app/api-request";
import { DATA_BOOKS, DATA_USER } from "@app/mocks/data";
import PreviewBook from "../page";

jest.mock("next-auth/react", () => ({
  useSession: jest.fn(),
}));

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@app/api-request", () => ({
  getBookById: jest.fn(),
  getUserById: jest.fn(),
  updateUserById: jest.fn(),
}));

jest.mock("@app/actions/auth", () => ({
  logout: jest.fn(),
}));

describe("Preview book Page", () => {
  (useSession as jest.Mock).mockReturnValue({
    data: {
      user: {
        isAdmin: true,
        email: "admin@gmail.com",
        id: "3733403",
        name: "admin",
        image: "https://i.ibb.co/RHMqQGr/man-1.png",
      },
      expires: "2024-12-31T23:59:59.999Z",
    },
    status: "authenticated",
  });

  beforeEach(() => {
    jest.clearAllMocks();
    (getBookById as jest.Mock).mockReturnValue({
      data: DATA_BOOKS[0],
    });
    (getUserById as jest.Mock).mockReturnValue({
      shelfBooks: DATA_USER[0].shelfBooks,
    });
  });

  it("Should render correctly snapshot", async () => {
    await act(async () => {
      const { container } = render(<PreviewBook params={{ id: "1" }} />);
      expect(container).toMatchSnapshot();
    });
  });
});
