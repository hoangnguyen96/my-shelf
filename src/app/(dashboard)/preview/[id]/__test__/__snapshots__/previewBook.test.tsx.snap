// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview book Page Should render correctly snapshot 1`] = `
<div>
  <div
    class="css-1ghqoze"
  >
    <a
      data-testid="click-back"
      href="#"
    >
      <svg
        class="chakra-icon css-1etxpkf"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          fill="currentColor"
        />
      </svg>
      <span
        class="chakra-text css-1mf4y6e"
      >
        Back to results
      </span>
    </a>
    <div
      class="css-199bpaw"
    >
      <div
        class="css-jxzlid"
      >
        <img
          alt="Preview book"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="278"
          src="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FLt7GmF2%2Fplace-holder.webp&w=640&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FLt7GmF2%2Fplace-holder.webp&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fi.ibb.co%2FLt7GmF2%2Fplace-holder.webp&w=640&q=75 2x"
          style="color: transparent; margin: 0px auto;"
          width="210"
        />
      </div>
      <div
        class="css-2j4sc6"
      >
        <p
          class="chakra-text css-0"
        />
        <p
          class="chakra-text css-1xo3mso"
        >
          By 
          ,
           
          <span
            class="chakra-text css-1xibt3d"
          />
        </p>
        <p
          class="chakra-text css-1xibt3d"
        >
           edition
        </p>
        <div
          class="css-1q1ptzx"
        >
          <div
            class="css-15zum9u"
          >
            <div
              class="css-k008qs"
            >
              <svg
                fill="#dfb300"
                height="18px"
                viewBox="0 0 24 24"
                width="18px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 2l3.09 6.26L22 9.27l-5 4.87L18.18 22 12 18.54 5.82 21 7 14.14 2 9.27l6.91-1.01L12 2z"
                />
              </svg>
              <svg
                fill="#dfb300"
                height="18px"
                viewBox="0 0 24 24"
                width="18px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 2l3.09 6.26L22 9.27l-5 4.87L18.18 22 12 18.54 5.82 21 7 14.14 2 9.27l6.91-1.01L12 2z"
                />
              </svg>
              <svg
                fill="#dfb300"
                height="18px"
                viewBox="0 0 24 24"
                width="18px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 2l3.09 6.26L22 9.27l-5 4.87L18.18 22 12 18.54 5.82 21 7 14.14 2 9.27l6.91-1.01L12 2z"
                />
              </svg>
              <svg
                fill="#dfb300"
                height="18px"
                viewBox="0 0 24 24"
                width="18px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 2l3.09 6.26L22 9.27l-5 4.87L18.18 22 12 18.54 5.82 21 7 14.14 2 9.27l6.91-1.01L12 2z"
                />
              </svg>
              <svg
                fill="#dfb300"
                height="18px"
                viewBox="0 0 24 24"
                width="18px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 2l3.09 6.26L22 9.27l-5 4.87L18.18 22 12 18.54 5.82 21 7 14.14 2 9.27l6.91-1.01L12 2z"
                />
              </svg>
            </div>
            <p
              class="chakra-text css-bg3st0"
            >
              <span
                class="chakra-text css-0"
              />
               Ratings
            </p>
          </div>
          <p
            class="chakra-text css-bg3st0"
          >
            <span
              class="chakra-text css-0"
            >
              25
            </span>
             Currently reading
          </p>
          <p
            class="chakra-text css-bg3st0"
          >
            <span
              class="chakra-text css-0"
            >
              119
            </span>
             Have read
          </p>
        </div>
        <div
          class="css-3jpwsf"
        >
          <div
            class="css-xbd6ua"
          >
            <p
              class="chakra-text css-182kmv3"
            >
              Availability
            </p>
            <div
              class="css-1c1kq07"
            >
              <svg
                fill="none"
                height="15px"
                viewBox="0 0 140 140"
                width="15px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M70 0C31.3317 0 0 31.3317 0 70C0 108.668 31.3317 140 70 140C108.668 140 140 108.668 140 70C140 31.3317 108.668 0 70 0ZM59.1971 94.8029C58.3894 95.6106 57.2452 96.2837 56.2356 96.2837C55.226 96.2837 54.0817 95.5769 53.2404 94.7692L34.3942 75.9231L40.3846 69.9327L56.2692 85.8173L98.2692 43.5144L104.159 49.6058L59.1971 94.8029Z"
                  fill="#4AC156"
                />
              </svg>
              <p
                class="chakra-text css-0"
              >
                Hard Copy
              </p>
            </div>
            <div
              class="css-1c1kq07"
            >
              <svg
                fill="none"
                height="15px"
                viewBox="0 0 140 140"
                width="15px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M70 0C31.3317 0 0 31.3317 0 70C0 108.668 31.3317 140 70 140C108.668 140 140 108.668 140 70C140 31.3317 108.668 0 70 0ZM59.1971 94.8029C58.3894 95.6106 57.2452 96.2837 56.2356 96.2837C55.226 96.2837 54.0817 95.5769 53.2404 94.7692L34.3942 75.9231L40.3846 69.9327L56.2692 85.8173L98.2692 43.5144L104.159 49.6058L59.1971 94.8029Z"
                  fill="#4AC156"
                />
              </svg>
              <p
                class="chakra-text css-0"
              >
                E - Book
              </p>
            </div>
            <div
              class="css-1c1kq07"
            >
              <svg
                fill="none"
                height="15px"
                viewBox="0 0 140 140"
                width="15px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M70 0C31.3317 0 0 31.3317 0 70C0 108.668 31.3317 140 70 140C108.668 140 140 108.668 140 70C140 31.3317 108.668 0 70 0ZM59.1971 94.8029C58.3894 95.6106 57.2452 96.2837 56.2356 96.2837C55.226 96.2837 54.0817 95.5769 53.2404 94.7692L34.3942 75.9231L40.3846 69.9327L56.2692 85.8173L98.2692 43.5144L104.159 49.6058L59.1971 94.8029Z"
                  fill="#4AC156"
                />
              </svg>
              <p
                class="chakra-text css-0"
              >
                Audio book
              </p>
            </div>
          </div>
          <div
            class="css-me0tyz"
          >
            <p
              class="chakra-text css-182kmv3"
            >
              Status
            </p>
            <div
              class="css-n0c263"
            >
              <p
                class="chakra-text css-1619v3u"
              >
                None
              </p>
            </div>
          </div>
        </div>
        <button
          class="chakra-button css-6vsr6c"
          type="button"
        >
          BORROW
        </button>
      </div>
      <div
        class="css-9wnjo"
      >
        <p
          class="chakra-text css-1d8s7zd"
        >
          About
           
          <span
            class="chakra-text css-ohxvsz"
          >
            Author
          </span>
        </p>
        <p
          class="chakra-text css-1uze92q"
        />
        <p
          class="chakra-text css-13pzcpe"
        />
        <img
          alt="Preview Author"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="100"
          src="/_next/image?url=%2Fimg.jpg&w=256&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=96&q=75 1x, /_next/image?url=%2Fimg.jpg&w=256&q=75 2x"
          style="color: transparent; position: absolute; right: 30%; top: 10%;"
          width="88"
        />
      </div>
    </div>
  </div>
</div>
`;
